### PR TITLE
Add inlined function decoding support for DWARF

### PR DESCRIPTION
**Pull request by danielocfb**
This change adds the necessary logic for decoding inlined functions when using the DWARF format. Like the rest of the code in the module, it's heavily based on the corresponding addr2line functionality.

Closes: #192

### CODE REVIEW

**Comment by danielocfb** [comment]
Will finish this up soon.

File: src/dwarf/function.rs [hunk 1]
```diff
@@ -112,6 +114,130 @@ where
...
+                Ordering::Less
+            } else if r1.call_depth > r2.call_depth {
+                Ordering::Greater
+            } else if r1.range.begin < r2.range.begin {
+                Ordering::Less
+            } else if r1.range.begin > r2.range.begin {
+                Ordering::Greater
+            } else {
+                Ordering::Equal
+            }
```
Comment (anakryiko) [inline src/dwarf/function.rs:192]: isn't there some way to use tuple here? like `(r1.call_depth, r1.range.begin).cmp((r2.call_depth, r2.range.begin))` ?

File: src/dwarf/function.rs [hunk 2]
```diff
@@ -112,6 +114,130 @@ where
...
+    pub(crate) fn parse(
+        dw_die_offset: gimli::UnitOffset<<R<'dwarf> as gimli::Reader>::Offset>,
+        unit: &gimli::Unit<R<'dwarf>>,
+        sections: &gimli::Dwarf<R<'dwarf>>,
+    ) -> Result<Self, Error> {
+        let mut entries = unit.entries_raw(Some(dw_die_offset))?;
+        let depth = entries.next_depth();
+        // The DIE offset we get is the one of the function. But we are
+        // interested in its children. So skip the necessary attributes.
+        let abbrev = entries.read_abbreviation().unwrap().unwrap();
```
Comment (anakryiko) [inline src/dwarf/function.rs:153]: is `.unwrap().unwrap()` safe to do here?

File: src/dwarf/units.rs [hunk 3]
```diff
@@ -299,15 +299,64 @@ impl<'dwarf> Units<'dwarf> {
...
+        Option<
+            impl ExactSizeIterator<
+                    Item = Result<(&'dwarf str, Option<Location<'slf>>), gimli::Error>,
+                > + 'slf,
+        >,
+        gimli::Error,
+    > {
+        for unit in self.find_units(probe) {
+            if let Some(function) = unit.find_function(probe, &self.dwarf)? {
+                let inlined_fns = function.parse_inlined_functions(unit.dw_unit(), &self.dwarf)?;
```
Comment (anakryiko) [inline src/dwarf/units.rs:324]: we don't cache this information, right? If that's the case, wouldn't it be more efficient to pass probe address to parse_inlined_functions() and skip parsing functions/ranges that definitely don't contain `probe`? If I read the current code correctly, we first parse everything, then filter out, which is less efficient.

File: src/dwarf/units.rs [hunk 4]
```diff
@@ -498,6 +509,37 @@ mod tests {
         });
     }
 
+    /// Benchmark the parsing of inlined function information, end-to-end.
```
Comment (anakryiko) [inline src/dwarf/units.rs:512]: please include your local benchmark results into the commit details when you are done, I'd be curious to see the relative performance. Thanks!

File: src/dwarf/function.rs (hunk 2 above)
Comment (d-e-s-o) [reply src/dwarf/function.rs:153]: It should be. At this point we have parsed the function already and we just redo these steps. Nevertheless, the read itself could probably short circuit more gracefully. Will adjust.

File: src/dwarf/function.rs (hunk 1 above)
Comment (d-e-s-o) [reply src/dwarf/function.rs:192]: Yes, the comparison can be shortened.

File: src/dwarf/units.rs (hunk 3 above)
Comment (d-e-s-o) [reply src/dwarf/units.rs:324]: We do cache the information.

File: src/dwarf/units.rs (hunk 4 above)
Comment (d-e-s-o) [reply src/dwarf/units.rs:512]: Will do.

**Comment by d-e-s-o** [comment]
> Will finish this up soon.

Should be ready now.

File: src/dwarf/units.rs (hunk 3 above)
Comment (anakryiko) [reply src/dwarf/units.rs:325]: ah, that's that LazyCell stuff, right?.. I see, disregard then.

**Review by anakryiko — APPROVED** [review]
Awesome, ship it!

File: src/dwarf/units.rs (hunk 3 above)
Comment (d-e-s-o) [reply src/dwarf/units.rs:325]: Yep.

**Pull request merged**